File: lib/test/t_MeasureEvaluation_std.cxx
#include <iostream>

#include "otrobopt/OTRobOpt.hxx"
#include "openturns/OT.hxx"


using namespace OT;
using namespace OTROBOPT;

int main()
{
  // First test: theta of dimension 1
  {
    Normal thetaDist(2.0, 0.1);
    Description input(2);
    input[0] = "x";
    input[1] = "theta";
    SymbolicFunction f_base(input, Description(1, "x*theta"));
    ParametricFunction f(f_base, Indices(1 , 1), Point(1, 1.0));

    Point x(1, 1.0);

    Collection <MeasureEvaluation> measures;
    measures.add(MeanMeasure(f, thetaDist));
    measures.add(VarianceMeasure(f, thetaDist));
    measures.add(WorstCaseMeasure(f, Uniform(-1.0, 4.0)));
    measures.add(WorstCaseMeasure(f, Uniform(-1.0, 4.0), false));
    measures.add(JointChanceMeasure(f, Normal(1.0, 1.0), GreaterOrEqual(), 0.95));
    measures.add(IndividualChanceMeasure(f, Normal(1.0, 1.0), GreaterOrEqual(), Point(1, 0.95)));
    measures.add(MeanStandardDeviationTradeoffMeasure(f, thetaDist, Point(1, 0.8)));
    measures.add(QuantileMeasure(f, thetaDist, 0.99));

    AggregatedMeasure aggregated(measures);
    measures.add(aggregated);

    const UnsignedInteger size = measures.getSize();
    for(UnsignedInteger i = 0; i < size; ++ i)
    {
      MeasureEvaluation measure(measures[i]);
      std::cout << measure << "(continuous)" << measure(x) << std::endl;
      {
        const UnsignedInteger N = 1000;
        LHSExperiment experiment(N);
        MeasureFactory factory(experiment);
        MeasureEvaluation discretizedMeasure(factory.build(measure));
        std::cout << discretizedMeasure << "(discretized LHS)" << discretizedMeasure(x) << std::endl;
      }
      {
        const UnsignedInteger N = 4;
        GaussProductExperiment experiment(Indices(1, N));
        MeasureFactory factory(experiment);
        MeasureEvaluation discretizedMeasure(factory.build(measure));
        std::cout << discretizedMeasure << "(discretized Gauss)" << discretizedMeasure(x) << std::endl;
      }
    }
  }
  // Second test: theta of dimension 2
  {
    Normal thetaDist(Point(2, 2.0), Point(2, 0.1), IdentityMatrix(2));
    Description input(3);
    input[0] = "x";
    input[1] = "theta0";
    input[2] = "theta1";
    SymbolicFunction f_base(input, Description(1, "x*theta0+theta1"));
    Indices indices(2);
    indices[0] = 1;
    indices[1] = 2;
    ParametricFunction f(f_base, indices, thetaDist.getMean());

    Point x(1, 1.0);

    Collection <MeasureEvaluation> measures;
    measures.add(MeanMeasure(f, thetaDist));
    measures.add(VarianceMeasure(f, thetaDist));
    measures.add(WorstCaseMeasure(f, JointDistribution(Collection<Distribution>(2, Uniform(-1.0, 4.0)))));
    measures.add(WorstCaseMeasure(f, JointDistribution(Collection<Distribution>(2, Uniform(-1.0, 4.0))), false));
    measures.add(JointChanceMeasure(f, thetaDist, GreaterOrEqual(), 0.5));
    measures.add(IndividualChanceMeasure(f, thetaDist, GreaterOrEqual(), Point(1, 0.5)));
    measures.add(MeanStandardDeviationTradeoffMeasure(f, thetaDist, Point(1, 0.8)));
    // FIXME: QuantileMeasure got slow with OT 1.2rc1, in the Brent step
    // measures.add(QuantileMeasure(f, thetaDist, 0.5));

    AggregatedMeasure aggregated(measures);
    measures.add(aggregated);

    const UnsignedInteger size = measures.getSize();
    for(UnsignedInteger i = 0; i < size; ++ i)
    {
      MeasureEvaluation measure(measures[i]);
      std::cout << measure << "(continuous)" << measure(x) << std::endl;
      {
        const UnsignedInteger N = 1000;
        LHSExperiment experiment(N);
        MeasureFactory factory(experiment);
        MeasureEvaluation discretizedMeasure(factory.build(measure));
        std::cout << discretizedMeasure << "(discretized LHS)" << discretizedMeasure(x) << std::endl;
      }
      {
        const UnsignedInteger N = 4;
        GaussProductExperiment experiment(measure.getDistribution(), Indices(2, N));
        MeasureFactory factory(experiment);
        MeasureEvaluation discretizedMeasure(factory.build(measure));
        std::cout << discretizedMeasure << "(discretized Gauss)" << discretizedMeasure(x) << std::endl;
      }
    }
  }
}

File: python/test/t_MeasureEvaluation_std.py
#!/usr/bin/env python

import openturns as ot
import openturns.testing as ott
import otrobopt

# First test: theta of dimension 1
thetaDist = ot.Normal(2.0, 0.1)
f_base = ot.SymbolicFunction(["x", "theta"], ["x*theta"])
f = ot.ParametricFunction(f_base, [1], [1.0])

x = [1.0]

measures = [
    otrobopt.MeanMeasure(f, thetaDist),
    otrobopt.VarianceMeasure(f, thetaDist),
    otrobopt.WorstCaseMeasure(f, ot.Uniform(-1.0, 4.0)),
    otrobopt.WorstCaseMeasure(f, ot.Uniform(-1.0, 4.0), False),
    otrobopt.JointChanceMeasure(f, ot.Normal(1.0, 1.0), ot.GreaterOrEqual(), 0.95),
    otrobopt.IndividualChanceMeasure(
        f, ot.Normal(1.0, 1.0), ot.GreaterOrEqual(), [0.95]
    ),
    otrobopt.MeanStandardDeviationTradeoffMeasure(f, thetaDist, [0.8]),
    otrobopt.QuantileMeasure(f, thetaDist, 0.99),
]
aggregated = otrobopt.AggregatedMeasure(measures)
measures.append(aggregated)

for i, measure in enumerate(measures):
    expectedDim = i if isinstance(measure, otrobopt.AggregatedMeasure) else 1
    ott.assert_almost_equal(measure(x).getDimension(), expectedDim)
    N = 10000
    experiment = ot.LHSExperiment(N)
    factory = otrobopt.MeasureFactory(experiment)
    discretizedMeasure = factory.build(measure)
    ott.assert_almost_equal(discretizedMeasure(x).getDimension(), expectedDim)
    N = 4
    experiment = ot.GaussProductExperiment([N])
    factory = otrobopt.MeasureFactory(experiment)
    discretizedMeasure = factory.build(measure)
    ott.assert_almost_equal(discretizedMeasure(x).getDimension(), expectedDim)

# Test known continuous values for dimension 1
ott.assert_almost_equal(measures[0](x), [2.0])  # MeanMeasure
ott.assert_almost_equal(measures[1](x), [0.01])  # VarianceMeasure
ott.assert_almost_equal(measures[2](x), [-1.0])  # WorstCase min
ott.assert_almost_equal(measures[3](x), [4.0])  # WorstCase max
ott.assert_almost_equal(measures[6](x), [0.48])  # MeanStdTradeoff

# Second test: theta of dimension 2
thetaDist = ot.Normal([2.0] * 2, [0.1] * 2, ot.IdentityMatrix(2))
f_base = ot.SymbolicFunction(["x", "theta0", "theta1"], ["x*theta0+theta1"])
f = ot.ParametricFunction(f_base, [1, 2], thetaDist.getMean())

x = [1.0]

measures = [
    otrobopt.MeanMeasure(f, thetaDist),
    otrobopt.VarianceMeasure(f, thetaDist),
    otrobopt.WorstCaseMeasure(f, ot.JointDistribution([ot.Uniform(-1.0, 4.0)] * 2)),
    otrobopt.WorstCaseMeasure(
        f, ot.JointDistribution([ot.Uniform(-1.0, 4.0)] * 2), False
    ),
    otrobopt.JointChanceMeasure(f, thetaDist, ot.GreaterOrEqual(), 0.95),
    otrobopt.IndividualChanceMeasure(f, thetaDist, ot.GreaterOrEqual(), [0.5]),
    otrobopt.MeanStandardDeviationTradeoffMeasure(f, thetaDist, [0.8]),
]
aggregated = otrobopt.AggregatedMeasure(measures)
measures.append(aggregated)

for i, measure in enumerate(measures):
    expectedDim = i if isinstance(measure, otrobopt.AggregatedMeasure) else 1
    ott.assert_almost_equal(measure(x).getDimension(), expectedDim)
    N = 10000
    experiment = ot.LHSExperiment(N)
    factory = otrobopt.MeasureFactory(experiment)
    discretizedMeasure = factory.build(measure)
    ott.assert_almost_equal(discretizedMeasure(x).getDimension(), expectedDim)
    N = 4
    experiment = ot.GaussProductExperiment(measure.getDistribution(), [N] * 2)
    factory = otrobopt.MeasureFactory(experiment)
    discretizedMeasure = factory.build(measure)
    ott.assert_almost_equal(discretizedMeasure(x).getDimension(), expectedDim)

# Test known continuous values for dimension 2
ott.assert_almost_equal(measures[0](x), [4.0])  # MeanMeasure: x*2.0+2.0 = 4.0
ott.assert_almost_equal(measures[1](x), [0.02])  # VarianceMeasure
ott.assert_almost_equal(measures[2](x), [-2.0])  # WorstCase min
ott.assert_almost_equal(measures[3](x), [8.0])  # WorstCase max
